### ISSUE TITLE
twisted.test.test_persisted.EphemeralTestCase.testEphemeral failing when there is no gc.get_referrers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3564]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)| @fijal reported|
|-|-|
|Trac ID|trac#3564|
|Type|defect|
|Created|2008-12-05 13:29:29Z|
|Branch|https://github.com/twisted/twisted/tree/optional-get-referrers-3564|

The problem is that __setstate__ uses get_referrers to show debugging info for logging
some information. Patch only logs this info when gc.get_referrers is there

Attachments:

* [ephemeral.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1f8/1f8f980df1092caab326d90f1b94810c153a1f5e/316729d2836471913627a9be13e6a75e39cb5133.diff) (631 bytes) - added by fijal on 2008-12-05 13:29:42Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3564 3564
type__defect defect
reporter__fijal fijal
priority__normal normal
milestone__ 
branch__branches_optional_get_referrers_3564 branches/optional-get-referrers-3564
branch_author__fijal fijal
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1228483769000000 1228483769000000
changetime__1228504628000000 1228504628000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)|@fijal removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [25568]) Branching to 'optional-get-referrers-3564'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [25569]) Apply ephemeral.diff
refs [#3564](https://github.com/twisted/twisted/issues/3564)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Looks good.  Fixes the test for this code on the PyPy builder.  I'll merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [25572]) Merge optional-get-referrers-3564

Author: fijal
Reviewer: exarkun
Fixes: [#3564](https://github.com/twisted/twisted/issues/3564)

Change `Ephemeral` to only use `gc.get_referrers` to report additional debugging
information if `gc.get_referrers` is actually available (runtimes such as PyPy
do not provide it).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
